### ISSUE TITLE
seport does not reassign port already used in system policy

### ISSUE THREAD

**Issue by Hubbitus** [ansible/ansible-modules-extras#2009]
##### Issue Type:
- Bug Report
##### Ansible Version:

```
ansible --version
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

I hope irrelevant
##### Environment:

Fedora 23 on my machine, CentOs 7 it is managed.
##### Summary:

seport module does not reassign port if it already assigned elsewhere. It report OK, but do nothing. Even no error.
##### Steps To Reproduce:

```
- seport: ports=8020 proto=tcp setype=http_port_t state=present
```

```
ok: [unidata.perf.client] => {"changed": false, "ports": ["8020"], "proto": "tcp", "setype": "http_port_t", "state": "present"}
```
##### Expected Results:

SELinux port 8020 assigned to http_port_t type.
##### Actual Results:

It does not happened:

```
# semanage port -l | ag '^http_port_t|8020'
hadoop_namenode_port_t         tcp      8020
http_port_t                    tcp      80, 81, 443, 488, 8008, 8009, 8443, 9000
```

I think next is a reason:

```
# semanage port -a -t http_port_t -p tcp 8020
ValueError: Port tcp/8020 already defined
```

But modify work:

```
# semanage port -m -t http_port_t -p tcp 8020
# semanage port -l | ag '^http_port_t|8020'
hadoop_namenode_port_t         tcp      8020
http_port_t                    tcp      8020, 80, 81, 443, 488, 8008, 8009, 8443, 9000
```

P.S. Initially reported to https://github.com/ansible/ansible/issues/14466

**Comment by mickael-ange** [comment]
Same problem with Ansible 2.0.2.0.
Can someone point to the `selinux` and `seobject` Python modules' documentation. I really cannot find it.
Is there an update function similar to `seport.update(port, proto, serange, setype)` ?
I'd like to try to fix this issue.
Cheers,

**Comment by kees-closed** [comment]
Same problem here on Fedora 23 and Fedora 24 both running Ansible 2.1.0.0.
I want to use 443 for ssh, but it can't because it's already in use by https. Ansible however doesn't give a warning and states in a green text that's it's okay, while it's not. As the bug reporter already pointed out, an add doesn't work if it's already assigned, a modify statement is needed in that case.

**Comment by avishefi** [comment]
+1
A flag to fail in case it already exists could help it determining whether it succeeds or not. Also, there should be an option to add the port to a rule without removing it from another one (semanage -m option does this right).

**Comment by ansibot** [comment]
@dankeder, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dankeder** [comment]
Hey, thanks for the bug report. It should be fixed by #2694.

For bot: resolved_by_pr #2694

**Comment by akrzos** [comment]
I am still seeing the issue with ansible 2.1.1.0 which what I believed the above p/r was merged into right? (as stated in the main comment of the P/R) or is there newer version of Ansible that p/r will be merged into?

``` shell
[akrzos@bithead ansible]$ ansible --version
ansible 2.1.1.0
  config file = /home/akrzos/code/browbeat-cloud07/browbeat/ansible/ansible.cfg
  configured module search path = Default w/o overrides
[akrzos@bithead ansible]$ ansible-playbook -i hosts install/browbeat.yml -vvvv -e 'browbeat_results_port=9001'
Using /home/akrzos/code/browbeat-cloud07/browbeat/ansible/ansible.cfg as config file
Loaded callback default of type stdout, v2.0

PLAYBOOK: browbeat.yml *********************************************************
1 plays in install/browbeat.yml

PLAY [undercloud] **************************************************************

TASK [browbeat : Allow httpd to listen to port (9001)] *************************
task path: /home/akrzos/code/browbeat-cloud07/browbeat/ansible/install/roles/browbeat/tasks/main.yml:231
<undercloud> ESTABLISH SSH CONNECTION FOR USER: stack
<undercloud> SSH: EXEC ssh -C -vvv -F /home/akrzos/code/browbeat-cloud07/ssh-config -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=stack -o ConnectTimeout=10 undercloud '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1471731783.35-191084968440005 `" && echo ansible-tmp-1471731783.35-191084968440005="` echo $HOME/.ansible/tmp/ansible-tmp-1471731783.35-191084968440005 `" ) && sleep 0'"'"''
<undercloud> PUT /tmp/tmpXahy0R TO /home/stack/.ansible/tmp/ansible-tmp-1471731783.35-191084968440005/seport
<undercloud> SSH: EXEC sftp -b - -C -vvv -F /home/akrzos/code/browbeat-cloud07/ssh-config -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=stack -o ConnectTimeout=10 '[undercloud]'
<undercloud> ESTABLISH SSH CONNECTION FOR USER: stack
<undercloud> SSH: EXEC ssh -C -vvv -F /home/akrzos/code/browbeat-cloud07/ssh-config -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=stack -o ConnectTimeout=10 -tt undercloud '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-xnzdezjugoenmpahdlghezgrtdykosss; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/stack/.ansible/tmp/ansible-tmp-1471731783.35-191084968440005/seport'"'"'"'"'"'"'"'"' && sleep 0'"'"''
ok: [undercloud] => {"changed": false, "invocation": {"module_args": {"ports": "9001", "proto": "tcp", "reload": true, "setype": "http_port_t", "state": "present"}, "module_name": "seport"}, "ports": ["9001"], "proto": "tcp", "setype": "http_port_t", "state": "present"}

PLAY RECAP *********************************************************************
undercloud                 : ok=1    changed=0    unreachable=0    failed=0   
```

And on my target machine:

``` shell
(root@xxxxxxxxxxxxxx) - (22:22) - (~)
-=>>semanage port --list | egrep "9001|^http_port_t"
http_port_t                    tcp      8088, 80, 81, 443, 488, 8008, 8009, 8443, 9000
tor_port_t                     tcp      6969, 9001, 9030, 9050, 9051, 9150
```

Note the `semanage` command can assign that port to the type I want:

``` shell
(root@xxxxxxxxxxxxxx) - (22:23) - (~)
-=>>semanage port -m -t http_port_t -p tcp 9001
Full path required for exclude: net:[4026532964].
Full path required for exclude: net:[4026532964].
(root@xxxxxxxxxxxxxx) - (22:24) - (~)
-=>>semanage port --list | egrep "9001|^http_port_t"
http_port_t                    tcp      9001, 8088, 80, 81, 443, 488, 8008, 8009, 8443, 9000
tor_port_t                     tcp      6969, 9001, 9030, 9050, 9051, 9150
```

**Comment by dankeder** [comment]
PR #2694  is not in 2.1.1.0 yet. The next version maybe?

**Comment by goldyfruit** [comment]
I'm using the master and I still having this issue.

**Comment by JoelFeiner** [comment]
Issue is still present on 2.3.0.

**Comment by dankeder** [comment]
At this moment the fix of this issue is merged only into devel. It's not in any of the stable-x.x branches (esp. not in stable-2.2). I'm not sure why :\.
